### PR TITLE
Implement Rust FFI wrapper of SMT C

### PR DESCRIPTION
**Pull request by joii2020**
Implement Rust FFI wrapper of SMT C
* Use "ckb-default-hash" at the low-level
* Implement  wrapper
* Add the Unit test about SMT C
* ckb_smt.c move to src/c

### CODE REVIEW

File: build.rs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+        .include("src/c")
+        .include("c/")
+        .include("../c/deps/ckb-c-stdlib")
+        .flag("-Wall")
+        .flag("-Werror")
+        .flag("-Wno-unused-parameter")
+        .flag("-Wno-nonnull")
+        .define("__SHARED_LIBRARY__", None)
+        .define("CKB_STDLIB_NO_SYSCALL_IMPL", None)
+        .compile("dl-c-impl");
```
Comment (XuJiandong) [inline build.rs:20]: Use another name: e.g. "smt-c-impl".

File: src/c/ckb_smt.c [hunk 2]
```diff
@@ -15,6 +15,14 @@ smt_state_t* smt_state_new(uint32_t capacity) {
   return state;
 }
 
+void smt_state_del(smt_state_t* state) {
```
Comment (XuJiandong) [inline src/c/ckb_smt.c:18]: I don't think we have used "malloc" "free" in SMT. If it's not used, just remove it.

File: src/ckb_smt.rs [hunk 3]
```diff
@@ -0,0 +1,109 @@
+use crate::collections::linked_list::LinkedList;
+use crate::H256;
+use core::{ffi::c_void, result::Result};
+
+#[link(name = "dl-c-impl", kind = "static")]
```
Comment (XuJiandong) [inline src/ckb_smt.rs:5]: choose another name instead of "dl-c-impl", e.g. "smt-c-impl" is OK.

**Review by XuJiandong — CHANGES_REQUESTED** [review]

File: src/ckb_smt.rs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    fn smt_state_normalize(state: *mut c_void);
+    fn smt_verify(
+        hash: *const u8,
+        state: *const c_void,
+        proof: *const u8,
+        proof_length: u32,
+    ) -> isize;
+}
+
+struct OriginCKBSmt {
```
Comment (xxuejie) [inline src/ckb_smt.rs:21]: Can we merge this struct with CkbSmtBuilder?

File: src/ckb_smt.rs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+            data: LinkedList::new(),
+        }
+    }
+
+    pub fn insert(&mut self, key: &H256, value: &H256) {
+        self.data.push_back((*key, *value));
+    }
+
+    pub fn build(&self) -> Result<CkbSmt, i32> {
+        let smt = CkbSmt {
```
Comment (xxuejie) [inline src/ckb_smt.rs:83]: What if we just use state of `*mut c_void` type here, and call all the C functions within this build Rust func?

Comment (xxuejie) [inline src/ckb_smt.rs:83]: What if we just use state of `*mut c_void` type here, and call all the C functions within this build Rust func?

File: src/ckb_smt.rs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+impl CkbSmtBuilder {
+    pub fn new() -> CkbSmtBuilder {
+        CkbSmtBuilder {
+            data: LinkedList::new(),
+        }
+    }
+
+    pub fn insert(&mut self, key: &H256, value: &H256) {
```
Comment (xxuejie) [inline src/ckb_smt.rs:78]: It's better we use `pub fn insert(mut self, key: &H256, value: &H256) -> self` as the interface the allow chaining calls.

File: src/ckb_smt.rs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        self.data.push_back((*key, *value));
+    }
+
+    pub fn build(&self) -> Result<CkbSmt, i32> {
```
Comment (xxuejie) [inline src/ckb_smt.rs:82]: Better to use `pub fn build(self) -> Result<CkbSmt, i32>`, so builder is properly discarded.

File: src/ckb_smt.rs (hunk 3 above)
Comment (joii2020) [reply src/ckb_smt.rs:5]: OK

File: c/rust-tests/src/tests/blake2b_decl_only.h [hunk 8]
```diff
@@ -21,9 +21,25 @@ typedef struct blake2b_state__ {
     uint8_t last_node;
 } blake2b_state;
 
+typedef struct blake2b_param__ {
```
Comment (XuJiandong) [inline c/rust-tests/src/tests/blake2b_decl_only.h:24]: If blake2b is used for testing, we can choose: https://crates.io/crates/blake2b-rs

**Review by XuJiandong — CHANGES_REQUESTED** [review]
Make sure the tests in "https://github.com/nervosnetwork/sparse-merkle-tree/tree/master/c/rust-tests/src/tests" is also migrated.

File: c/rust-tests/src/tests/blake2b_decl_only.h (hunk 8 above)
Comment (joii2020) [reply c/rust-tests/src/tests/blake2b_decl_only.h:24]: Yes, Low-level uses the blake2b-rs.
Now I just wrote a few simple use unit test, I want to wait until the interface is fixed before I complete all of them.

**Review by XuJiandong — APPROVED** [review]

**Pull request merged**